dwave/optimization/src/simplex.cpp

```diff
@@ -0,0 +1,563 @@
+
+    Matrix(std::vector<double>&& data, ssize_t n, ssize_t m)
+            : n_(n), m_(m), buffer_(std::move(data)) {
+        assert(n_ * m_ == static_cast<ssize_t>(buffer_.size()) &&
+               "shape does not match provided data");
+    }
+
+    Matrix(std::span<const double> data, ssize_t n, ssize_t m)
+            : Matrix(std::vector<double>(data.begin(), data.end()), n, m) {}
+
+    double& operator()(ssize_t i, ssize_t j) {
+        if (i < 0) i += n_;
+        if (j < 0) j += m_;
+        assert(i >= 0 && i < n_);
+        assert(j >= 0 && j < m_);
+        return buffer_[i * m_ + j];
+    }
+
+    const double& operator()(ssize_t i, ssize_t j) const {
+        if (i < 0) i += n_;
+        if (j < 0) j += m_;
+        assert(i >= 0 && i < n_);
+        assert(j >= 0 && j < m_);
+        return buffer_[i * m_ + j];
+    }
+
+    ssize_t n() const { return n_; }
+
+    ssize_t m() const { return m_; }
+
+ private:
+    ssize_t n_;
+    ssize_t m_;
+
+    std::vector<double> buffer_;
+};
+
+/// Find the pivot column. Will return -1 if no candidate column is found.
+ssize_t _pivot_col(Matrix& T, double tolerance, bool bland) {
+    double min_val = std::numeric_limits<double>::infinity();
+    ssize_t col = -1;
+    for (ssize_t j = 0; j < T.m() - 1; j++) {
+        double val = T(-1, j);
+        if (val < -tolerance) {
+            // Bland's rule: return the first index with negative value
+            if (bland) return j;
+
+            if (val < min_val) {
+                min_val = val;
+                col = j;
+            }
+        }
+    }
+    return col;
+}
+
+/// Find the pivot row. Will return -1 if no candidate row is found.
+ssize_t _pivot_row(Matrix& T, const std::vector<ssize_t>& basis, ssize_t pivcol, ssize_t phase,
+                   double tolerance, bool bland) {
+    ssize_t k = phase == 1 ? 2 : 1;
+
+    double min_q = std::numeric_limits<double>::infinity();
+    std::vector<ssize_t> min_rows;
+    for (ssize_t i = 0; i < T.n() - k; i++) {
+        if (T(i, pivcol) <= tolerance) continue;
+
+        double q = T(i, -1) / T(i, pivcol);
+        if (q < min_q) {
+            min_rows.clear();
+            min_rows.push_back(i);
+            min_q = q;
+        } else if (q == min_q) {
+            min_rows.push_back(i);
+        }
+    }
+
+    if (min_rows.size() == 0) return -1;
+
+    if (!bland) return min_rows[0];
+
+    // Bland's rule
+    ssize_t min_basis = std::numeric_limits<ssize_t>::max();
+    ssize_t min_row = -1;
+    for (const ssize_t& row : min_rows) {
+        if (basis[row] < min_basis) {
+            min_basis = basis[row];
+            min_row = row;
+        }
+    }
+    return min_row;
+}
+
+void _apply_pivot(Matrix& T, std::vector<ssize_t>& basis, ssize_t pivrow, ssize_t pivcol,
+                  double tolerance) {
+    basis[pivrow] = pivcol;
+
+    double pivval = T(pivrow, pivcol);
+
+    for (ssize_t j = 0; j < T.m(); j++) {
+        T(pivrow, j) /= pivval;
```

Comment:
If T.m() can be large, might be best to store the reciprocal of pivval outside the loop and multiply by it.

Reply:
Actually, given that this is more of a reference implementation and not focused on performance in any way, I'm inclined to leave it as division. The other upside of using actual division is that the floating point error remains smaller, which may help with numerical stability of the overall algorithm.